util/src/main/java/tech/pegasys/teku/util/async/SafeFuture.java

```diff
@@ -332,4 +374,84 @@ public void finish(final Consumer<T> onSuccess, final Consumer<Throwable> onErro
   public SafeFuture<T> orTimeout(final long timeout, final TimeUnit unit) {
     return (SafeFuture<T>) super.orTimeout(timeout, unit);
   }
+
+  @SafeVarargs
+  @SuppressWarnings("unchecked")
+  public final SafeFuture<T> or(SafeFuture<T>... others) {
+    SafeFuture<T>[] futures = Arrays.copyOf(others, others.length + 1);
+    futures[others.length] = this;
+    return anyOf(futures).thenApply(o -> (T) o);
+  }
+
+  /**
+   * Derives a {@link SafeFuture} which yields the same result as this {@link SafeFuture} if no
+   * {@link Interruptor} was triggered before this future is done.
+   *
+   * <p>If any of supplied {@link Interruptor}s is triggered the returned {@link SafeFuture} is
+   * completed exceptionally. The exception thrown depends on which specific Interruptor was
+   * triggered
+   *
+   * <p>The key feature of this method is that {@code interruptFuture} contained in Interruptor
+   * doesn't hold the reference to dependent futures after they complete. It's desired to consider
+   * this for long living interrupting futures to avoid memory leaks
+   *
+   * @param interruptors a set of interruptors which futures trigger interruption if complete
+   *     (normally or exceptionally)
+   * @see #createInterruptor(CompletableFuture, Supplier)
+   */
+  public SafeFuture<T> orInterrupt(Interruptor... interruptors) {
+    CompletableFuture<?>[] allFuts = new CompletableFuture<?>[interruptors.length + 1];
+    allFuts[0] = this;
+    for (int i = 0; i < interruptors.length; i++) {
+      allFuts[i + 1] = interruptors[i].interruptFuture;
+    }
+    SafeFuture<T> ret = new SafeFuture<>();
+    anyOf(allFuts)
+        .whenComplete(
+            (res, err) -> {
+              if (this.isDone()) {
+                this.propagateTo(ret);
+              } else {
+                for (Interruptor interruptor : interruptors) {
+                  if (interruptor.interruptFuture.isDone()) {
+                    try {
+                      interruptor.getInterruptFuture().get();
+                      ret.completeExceptionally(interruptor.getExceptionSupplier().get());
+                    } catch (Exception e) {
+                      ret.completeExceptionally(e);
+                    }
+                  }
+                }
+              }
+            })
+        .exceptionally(e -> null); // all exceptions are propagated to ret
```

Comment:
Given this should never happen, could we log an error if it does instead of ignoring?

Reply:
This is similar to `SafeFuture.propagate` in the sense that any exception is propagated to the returned future. So if `anyOf` completes exceptionally this is a regular situation on this level. 
Same as `SafeFuture.propagate` added `SuppressWarning` with explanation comment